katdal/van_vleck.py

```diff
@@ -0,0 +1,83 @@
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Routines for performing quantisation (Van Vleck) correction."""
+
+from __future__ import print_function, division, absolute_import
+
+import math
+
+import numpy as np
+import numba
+
+
+@numba.vectorize(['f8(f8, f8)'], nopython=True, cache=True)
+def norm0_cdf(x, scale):
+    """Fast zero-mean (loc=0) implementation of :meth:`scipy.stats.norm.cdf`."""
+    return 0.5 * (math.erf(np.sqrt(0.5) * x / scale) + 1.)
+
+
+def _quant_norm0_pmf(levels, var=1.0):
+    """Probability mass function of quantised zero-mean normal variable."""
+    edges = np.r_[-np.inf, levels[:-1] + np.diff(levels) / 2., np.inf]
+    return np.diff(norm0_cdf(edges, np.sqrt(var)))
+
+
+def _squared_quant_norm0_mean(levels, var=1.0):
+    """Mean of squared quantised zero-mean normal variable (same shape as `var`)."""
+    levels = np.asarray(levels)
+    # Allow var and levels to be broadcast against each other, with levels as last dimension
+    var = np.asarray(var)[..., np.newaxis]
+    pmf = _quant_norm0_pmf(levels, var)
+    return pmf.dot(levels * levels)
+
+
+def autocorr_lookup_table(levels, size=4000):
+    """Lookup table that corrects complex autocorrelation quantisation effects.
+
+    This maps the variance of a quantised complex voltage signal to the variance
+    of the unquantised signal under the assumption that the signal is proper
+    (circularly-symmetric) complex normally distributed.
+
+    Parameters
+    ----------
+    levels : sequence of float
+        Quantisation levels for real and imaginary components of voltage signal
+    size : int, optional
+        Size of lookup table
+
+    Returns
+    -------
+    quantised_autocorr_table, true_autocorr_table : array of float, shape (`size`,)
+        Lookup table associating quantised autocorrelations and unquantised
+        autocorrelations (ie power/variance of complex signals)
+    """
+    # Terminology:
+    # x = Proper complex normal voltage signal (zero-mean)
+    # rxx = Power (variance) *per* real/imag component of unquantised / true x
+    # sxx = Power (variance) *per* real/imag component of quantised x
+    sxx_max = np.abs(levels).max() ** 2
+    # Sweep across range of true power values, placing more table entries at tricky lower end
+    rxx_grid = np.r_[np.logspace(-2.4, 0, size // 2, endpoint=False),
+                     np.logspace(0, np.log10(sxx_max) + 8, size - 2 - size // 2)]
```

Comment:
You're right. I like that suggestion!

Reply:
The change now shifts the top as well. Perhaps `np.log10(sxx_max)` should become `np.log10(sxx_max / sxx_min)`?

Reply:
The alternative was plonking `log10(sxx_min)` into three places, so I just accepted that the fourth one could change too, since we are wayyy above sxx_max already (8 orders of magnitude!).

But I do like your suggestion... Especially if I ever tweak the top down to around the realistic maximum rxx.

Reply:
Implemented your suggestion.